File: leetcode/length-of-longest-substring.ts
function lengthOfLongestSubstring(s: string): number {
  let maxLength = 0,
    start = 0;
  const charIndexMap = new Map<string, number>();

  for (let i = 0; i < s.length; i++) {
    if (charIndexMap.has(s[i]) && charIndexMap.get(s[i])! >= start) {
      start = charIndexMap.get(s[i])! + 1;
    }

    charIndexMap.set(s[i], i);
    maxLength = Math.max(maxLength, i - start + 1);
  }

  return maxLength;
}

File: leetcode/length-of-longest-substring.go
package leetcode

func lengthOfLongestSubstring(s string) int {
	maxLength, start := 0, 0
	charIndexMap := make(map[byte]int)
	for i := 0; i < len(s); i++ {
		if index, ok := charIndexMap[s[i]]; ok && index >= start {
			start = index + 1
		}
		charIndexMap[s[i]] = i
		if currentLength := i - start + 1; currentLength > maxLength {
			maxLength = currentLength
		}
	}
	return maxLength
}
